nexus/types/src/deployment.rs

```diff
@@ -0,0 +1,528 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Types representing deployed software and configuration
+//!
+//! For more on this, see the crate-level documentation for `nexus/deployment`.
+//!
+//! This lives in nexus/types because it's used by both nexus/db-model and
+//! nexus/deployment.  (It could as well just live in nexus/db-model, but
+//! nexus/deployment does not currently know about nexus/db-model and it's
+//! convenient to separate these concerns.)
+
+pub use crate::inventory::OmicronZoneConfig;
+pub use crate::inventory::OmicronZoneDataset;
+pub use crate::inventory::OmicronZoneType;
+pub use crate::inventory::OmicronZonesConfig;
+pub use crate::inventory::ZpoolName;
+use omicron_common::address::Ipv6Subnet;
+use omicron_common::address::SLED_PREFIX;
+use omicron_common::api::external::Generation;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use uuid::Uuid;
+
+/// Fleet-wide deployment policy
+///
+/// The **policy** represents the deployment controls that people (operators and
+/// support engineers) can modify directly under normal operation.  In the
+/// limit, this would include things like: which sleds are supposed to be part
+/// of the system, how many CockroachDB nodes should be part of the cluster,
+/// what system version the system should be running, etc.  It would _not_
+/// include things like which services should be running on which sleds or which
+/// host OS version should be on each sled because that's up to the control
+/// plane to decide.  (To be clear, the intent is that for extenuating
+/// circumstances, people could exercise control over such things, but that
+/// would not be part of normal operation.)
+///
+/// The current policy is pretty limited.  It's aimed primarily at supporting
+/// the add/remove sled use case.
+pub struct Policy {
+    /// set of sleds that are supposed to be part of the control plane, along
+    /// with information about resources available to the planner
+    pub sleds: BTreeMap<Uuid, SledResources>,
+}
+
+/// Describes the resources available on each sled for the planner
+pub struct SledResources {
+    /// zpools on this sled
+    ///
+    /// (used to allocate storage for control plane zones with persistent
+    /// storage)
+    pub zpools: BTreeSet<ZpoolName>,
+
+    /// the IPv6 subnet of this sled on the underlay network
+    ///
+    /// (implicitly specifies the whole range of addresses that the planner can
+    /// use for control plane components)
+    pub subnet: Ipv6Subnet<SLED_PREFIX>,
+}
+
+/// Describes a complete set of software and configuration for the system
+// Blueprints are a fundamental part of how the system modifies itself.  Each
```

Comment:
I had that previously.  But this is a publicly-exposed type, so the block comment is used for API consumers, not just Rust consumers.  I remember discussing this months ago and deciding that unless there was a really compelling reason to try to support separate API docs and rustdoc for the same type, we'd just let the rustdoc reflect API docs.

Reply:
Hmm, but this is an API type for the _internal_ API, right? Do those exist? (To be fair, another question is whether we generate rustdocs for crates. I've done that locally occasionally but it's not a habit.)

Reply:
I've been heavily using API docs for internal APIs (generated by hand with redoc).  Then again I also use rustdocs via rust-analyzer.  But I'm inclined to treat both APIs the same here -- part of the idea behind using OpenAPI everywhere was so that we could have a good experience using our own APIs through the patterns/processes/tooling we use for the external API.  I find it annoying when I'm using the internal API docs and I get a lot of irrelevant Rust-isms.

Reply:
Fair enough. I don't feel super strongly about this either way, happy to defer to you.